File: js/task3/S3_PIPE_02.js
function compose(...fns) {
  return function (value) {
    let res = value;

    for (let i = fns.length - 1; i >= 0; i--) {
      res = fns[i](res);
    }

    return res;
  };
}

// Tests
const add1 = (x) => x + 1;
const double = (x) => x * 2;

const run = compose(add1, double);
console.log(run(3)); // 7

File: python/stage3/S3_PIPE_02.py
def compose(*fns):
    def apply(x):
        for fn in reversed(fns):
            x = fn(x)
        return x
    return apply


# Tests
f_pipe = pipe(lambda x: x + 1, lambda x: x * 2)
f_comp = compose(lambda x: x + 1, lambda x: x * 2)

print(f_pipe(3))
print(f_comp(3))
